mlx/backend/common/conv.cpp

```diff
@@ -1000,6 +1001,30 @@ void explicit_gemm_conv_ND_cpu(
     copy(wt, gemm_wt, ctype);
   }
 
+  if (flip) {
+    auto gemm_wt_ = array(gemm_wt.shape(), float32, nullptr, {});
+    copy(gemm_wt, gemm_wt_, CopyType::Vector);
+
+    const int ndim = wt.ndim();
+    const int N = wt.shape(0);
+    const int C = wt.shape(ndim - 1);
+
+    // Calculate the total size of the spatial dimensions
+    int spatial_size = 1;
+    for (int d = 1; d < ndim - 1; ++d) {
+      spatial_size *= wt.shape(d);
+    }
+
+    // Use vDSP to reverse the spatial dimensions for each N,C combination
+    for (int n = 0; n < N; ++n) {
+      for (int c = 0; c < C; ++c) {
+        float* src = gemm_wt_.data<float>() + (n * spatial_size * C + c);
+        vDSP_vrvrs(src, C, spatial_size);
```

Comment:
Unfortunately we can't guarantee that we 'll have accelerate here since it is the common CPU backend. I think doing it with `copy_inplace` would need another allocation so we may need a `flip` function. Something as simple as

```cpp
template <typename T>
void flip_spatial_dims(T *x, size_t batch, size_t spatial_size, size_t dims) {
  for (size_t i=0; i<batch; i++) {
    T *top = x + i * spatial_size * dims;
    T *bottom = x + i * spatial_size * dims + (spatial_size-1) * dims;
    for (size_t j=0; j<spatial_size/2; j++) {
      for (size_t k=0; k<dims; k++) {
        std::swap(top[k], bottom[k]);
      }
      top += dims;
      bottom -= dims;
    }
  }
}
```

seems to work well without any significant slow down on my M2 air (it is a bit slower than accelerate I do believe but the bulk of the time is spent on the matmul anyway). Feel free to benchmark yourself or try something else. The code above is literally the simplest way you can write this.

Reply:
Thanks for the clarification. I thought `Accelerate` would be fine as we are using `cblas_sgemm` from it already, but I guess Linux users could just use it from LAPACK, which would not include `vDSP_vrvrs `.